### ISSUE TITLE
device tracker: use HA zones instead of zones defined in the whistle app?

### ISSUE THREAD

**Issue by dbdmdbdmdbdm** [RobertD502/home-assistant-whistle#9]
Instead of using the zones defined in the whistle app, can device tracker instead set the state based on where it actually is? That way the HA zones will be consistent with other device_trackers in HA and the user wouldnt have to match zones in whistle app with zones in HA.

**Comment by dbdmdbdmdbdm** [comment]
I've also noticed that the "Away" state doesnt seem to be updated despite the device tracker gps coordinates being out of any whistle (or HA) zones.

**Comment by RobertD502** [comment]
Comment out lines 132-143 within the device_tracker.py file if you want to rely on Home Assistant zones.

**Comment by dbdmdbdmdbdm** [comment]
I commented them out and the device_tracker entity is no longer working.

**Comment by RobertD502** [comment]
Please provide screenshots of what you've done as the location name property isn't a requirement for device tracker entities, so, commenting that out should never break the entity.

**Comment by dbdmdbdmdbdm** [comment]
This seems to be the error
`Logger: homeassistant.components.device_tracker
Source: util/location.py:95
Integration: Device tracker ([documentation](https://www.home-assistant.io/integrations/device_tracker), [issues](https://github.com/home-assistant/home-assistant/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+device_tracker%22))
First occurred: 11:05:09 AM (2 occurrences)
Last logged: 11:05:10 AM

Error adding entities for domain device_tracker with platform whistle
Error while setting up whistle platform for device_tracker
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 428, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 673, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 777, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 533, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 571, in _async_write_ha_state
    state = self._stringify_state(available)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 539, in _stringify_state
    if (state := self.state) is None:
  File "/usr/src/homeassistant/homeassistant/components/device_tracker/config_entry.py", line 249, in state
    zone_state = zone.async_active_zone(
  File "/usr/src/homeassistant/homeassistant/components/zone/__init__.py", line 121, in async_active_zone
    zone_dist = distance(
  File "/usr/src/homeassistant/homeassistant/util/location.py", line 70, in distance
    result = vincenty((lat1, lon1), (lat2, lon2))
  File "/usr/src/homeassistant/homeassistant/util/location.py", line 95, in vincenty
    U1 = math.atan((1 - FLATTENING) * math.tan(math.radians(point1[0])))
TypeError: must be real number, not str`

This is what i commented out
`#    @property removed below lines to support this https://github.com/RobertD502/home-assistant-whistle/issues/9
#    def location_name(self) -> str:
#        """Returns the Whistle location, as defined in the app.
#        If the tracker is not in a pre-defined location, location
#        of Away is returned.
#        """

#        if self.pet_data.data['last_location']['place']['id']:
#            location_id = self.pet_data.data['last_location']['place']['id']
#            return self.location_dict.get(location_id)
#        else:
#            return "Away"
`

**Comment by RobertD502** [comment]
So, the Home Assistant Developer documentation is wrong and states that the latitude and longitude should be returned as strings instead of integers. Seems that is completely wrong and explains the TypeError that you're reporting. I'll have to bring this up with them to fix their documentation.

To get the device tracker working, replace the code inside of the device_tracker.py file with this:

```python
""" Device Tracker platform for Whistle integration."""
from __future__ import annotations

from typing import Any

from whistleaio.model import Pet

from homeassistant.components.device_tracker.const import SourceType
from homeassistant.components.device_tracker.config_entry import TrackerEntity
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN
from .coordinator import WhistleDataUpdateCoordinator

async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    """ Set Up Whistle Device Tracker Entities. """

    coordinator: WhistleDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]

    device_trackers = []


    for pet_id, pet_data in coordinator.data.pets.items():

            """ Device Trackers """
            if pet_data.data['device'] and pet_data.data['device']['has_gps']:
                device_trackers.extend((
                        WhistleTracker(coordinator, pet_id),
                    ))
    async_add_entities(device_trackers)

class WhistleTracker(CoordinatorEntity, TrackerEntity):
    """ Representation of Whistle GPS Tracker. """

    def __init__(self, coordinator, pet_id):
        super().__init__(coordinator)
        self.pet_id = pet_id


    @property
    def pet_data(self) -> Pet:
        """ Handle coordinator pet data. """

        return self.coordinator.data.pets[self.pet_id]

    @property
    def location_dict(self) -> dict[int, str]:
        """ Create a dictionary for all pre-defined Whistle
        locations as a dict of location id: location name.
        """

        locations: dict[int, str] = {}
        for place in self.pet_data.places:
            locations[place['id']] = place['name']

        return locations

    @property
    def device_info(self) -> dict[str, Any]:
        """ Return device registry information for this entity. """

        return {
            "identifiers": {(DOMAIN, self.pet_data.id)},
            "name": self.pet_data.data['name'],
            "manufacturer": "Whistle",
            "model": self.pet_data.data['device']['model_id'],
            "configuration_url": "https://www.whistle.com/",
        }

    @property
    def unique_id(self) -> str:
        """ Sets unique ID for this entity. """

        return str(self.pet_data.id) + '_tracker'

    @property
    def name(self) -> str:
        """ Return name of the entity. """

        return "Whistle tracker"

    @property
    def has_entity_name(self) -> bool:
        """ Indicate that entity has name defined. """

        return True

    @property
    def icon(self):
        """ Determine what icon to use. """

        if self.pet_data.data['profile']['species'] == 'dog':
            return 'mdi:dog'
        if self.pet_data.data['profile']['species'] == 'cat':
            return 'mdi:cat'

    @property
    def source_type(self) -> SourceType:
        """ Return GPS as the source. """

        return SourceType.GPS

    @property
    def latitude(self) -> int:
        """ Return most recent latitude. """

        return self.pet_data.data['last_location']['latitude']

    @property
    def longitude(self) -> int:
        """ Return most recent longitude. """

        return self.pet_data.data['last_location']['longitude']

    @property
    def battery_level(self) -> int:
        """ Return tracker current battery percent. """

        return self.pet_data.data['device']['battery_level']

    @property
    def location_accuracy(self) -> int:
        """ Return last location gps accuracy. """

        return int(self.pet_data.data['last_location']['uncertainty_meters'])

```

**Comment by dbdmdbdmdbdm** [comment]
Device tracker is now working, will let you know if it updates status once i go on a walk! Thank you Rob!

**Comment by RobertD502** [comment]
No problem!

**Comment by RobertD502** [comment]
@dbdmdbdmdbdm I realized that using this as a solution would be a PITA for you as you'd have to make this change after every update that I push out. I just released version `0.1.2` which allows you to choose if you want to use Whistle Zones or Home Assistant Zones. By default, Whistle zones are used. So, after updating to 0.1.2, click on configure and then select Home Assistant as your option to use zones that you defined within Home Assistant:

<img width="275" alt="image" src="https://user-images.githubusercontent.com/52541649/190012105-f40bbda5-7ce3-405d-ac05-6348e736a047.png">
<img width="393" alt="image" src="https://user-images.githubusercontent.com/52541649/190012196-8db749df-9d37-40d4-934f-b7279e9b06f1.png">

Also, thanks for bringing to my attention that the device tracker wasn't updating to Away when out of a Whistle zone. Further investigation revealed that Whistle servers return a place id in the "last location" even if the pet isn't actually at that location. They do set the status to "outside_geofence_range" so I'm using that now to correctly set the status to Away.
